x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sample/SampleIterator.java

```diff
@@ -0,0 +1,327 @@
+import org.elasticsearch.xpack.eql.session.Payload.Type;
+import org.elasticsearch.xpack.ql.util.ActionListeners;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Stack;
+
+import static org.elasticsearch.action.ActionListener.runAfter;
+import static org.elasticsearch.action.ActionListener.wrap;
+import static org.elasticsearch.common.Strings.EMPTY_ARRAY;
+import static org.elasticsearch.xpack.eql.execution.assembler.SampleQueryRequest.COMPOSITE_AGG_NAME;
+import static org.elasticsearch.xpack.eql.execution.search.RuntimeUtils.prepareRequest;
+
+public class SampleIterator implements Executable {
+
+    private final Logger log = LogManager.getLogger(SampleIterator.class);
+
+    private final QueryClient client;
+    private final List<SampleCriterion> criteria;
+    private final Stack<Page> stack = new Stack<>();
+    private final int maxStages;
+    private final List<Sample> samples;
+    private final int fetchSize;
+
+    private long startTime;
+
+    public SampleIterator(QueryClient client, List<SampleCriterion> criteria, int fetchSize) {
+        this.client = client;
+        this.criteria = criteria;
+        this.maxStages = criteria.size();
+        this.fetchSize = fetchSize;
+        this.samples = new ArrayList<>();
+    }
+
+    @Override
+    public void execute(ActionListener<Payload> listener) {
+        startTime = System.currentTimeMillis();
+        // clear the memory at the end of the algorithm
+        advance(runAfter(listener, () -> {
+            stack.clear();
+            samples.clear();
+            client.close(listener.delegateFailure((l, r) -> {}));
+        }));
+    }
+
+    /*
+     * Starting point of the iterator, which also goes through all the criterions and gathers initial results pages.
+     */
+    private void advance(ActionListener<Payload> listener) {
+        int currentStage = stack.size();
+        if (currentStage < maxStages) {
+            // excessive logging for testing purposes
+            log.trace("Advancing from stage [{}]", currentStage);
+            SampleCriterion criterion = criteria.get(currentStage);
+            SampleQueryRequest request;
+
+            // incorporate the previous stage composite keys in the current stage query
+            if (currentStage > 0) {
+                request = criterion.midQuery();
+                Page previousResults = stack.peek();
+                SampleCriterion previousCriterion = criteria.get(currentStage - 1);
+                request.multipleKeyPairs(previousCriterion.keys(previousResults.hits), previousResults.keys);
+            } else {
+                request = criterion.firstQuery();
+            }
+
+            final SampleQueryRequest rr = request;
+            log.trace("Querying stage [{}] {}", currentStage, request);
+            client.query(request, wrap(r -> {
+                Aggregation a = r.getAggregations().get(COMPOSITE_AGG_NAME);
```

Comment:
If I'm not wrong, this logic is exactly the same as `nextPage()` (apart from the log message), probably it's worth extracting it as a method and use it in both contexts